C# encode to a given array

ProgrammerAl opened the pull request:
This is the implementation for what was discussed in: #144 

Sorry this took me longer to get to than I had hoped. The final result was a little different than what I original speced out. Here's what's implemented:

New overloads for `BebopWriter.Create()`
- `BebopWriter.Create(int initialCapacity)`
  - Creates a new `byte[]` buffer with a size equal to `initialCapacity`
- `BebopWriter.Create(byte[] buffer)`
  - Uses the input `buffer` when encoding

C# Compiler creates a new `EncodeIntoBuffer(byte[] outBuffer)` method for each record (along with a static version like the normal Encode() method)
- `EncodeIntoBuffer()` is given a `byte[]` to write out the seralized record into. If `outBuffer` is not big enough to hold the final serialized message, it is expanded like a normal call to `Encode()`. ie a new `byte[]` object is allocated and copied to.
- Return type is different from `Encode()`. Returns instance of `readonly struct BebopWriteResult`
- `BebopWriteResult` has 3 properties
  - Result
    - Enum of type `WriteResult`
    - Value will either be `Success` or `SuccessButGrewBuffer`
  - Written
    - How many bytes were written to the buffer
  - Buffer
    - Reference the the buffer with the final serialized message
    - If encoded into a given buffer, and the entire message fit into that buffer, this is the same reference to the input buffer
    - If the message did not fit into the original buffer and it was grown, ie a new `byte[]` object was allocated, `Buffer` is a reference to that new array


There are also some Benchmarks added to show the new functionality.

AndrewMD5 commented on Runtime/C#/Runtime/BebopWriteResult.cs:
```diff
@@ -0,0 +1,37 @@
+        /// <summary>
+        /// The final instance of the buffer written to. If Bebop Writer was given a buffer to start with, but had to grow it to fit the fill serialized message, this will be a different instance than what was passed in.
+        /// </summary>
+        public byte[] Buffer { get; }
```
A few things:

1) [Properties should not return arrays](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1819). 

2) If `BebopWriter` is given a buffer that is too small to contain the encoded record, then we should treat this as an error and return `BufferTooSmall` rather than attempting to grow the array and returning `SuccessButGrewBuffer`. It is the responsibility of developers to ensure their buffers are the right size. 

With that change you can remove the `Buffer` property from this struct and upon a successful encoding the buffer that was passed into `EncodeIntoBuffer` will be the same one that contains the encoded bytes.

ProgrammerAl replied:
I prefer what you suggested, but this one will be difficult. Since `void BebopWriter.GrowBy()` is called each time a field is written to the buffer, even if it doesn't need to grow, I'm worried we'd have to change it to return something like `EncodeResult` on each call, then add an if check around it to know if we should keep encoding, or return immediatly.

How do you feel about throwing an exception inside `BebopWriter.GrowBy()` when that instance of `BebopWriter` can't grow an input buffer? I haven't benchmarked it yet, but here's what I'm thinking it would end up looking like:
```cs
private void GrowBy(int amount)
{
    if ((Length & 0xC0000000) != 0)
    {
        throw new BebopViewException("A Bebop View cannot grow beyond 2 gigabytes.");
    }
    if (Length + amount > _buffer.Length)
    {
        // START NEW CODE
        if (_isInputBuffer) // Note: _isInputBuffer is set by the constructor
        {
            throw new BebopViewException("A Bebop View cannot grow a buffer that was input during View creation.");
        }
        // END NEW CODE

        //snip: grow buffer
    }
    Length += amount;
}
```

ProgrammerAl replied:
I ended up implementing something very similar to this approach. Ran the same benchmarks and didn't notice any performance regression.

AndrewMD5 commented on Runtime/C#/Runtime/BebopWriteResult.cs:
```diff
@@ -0,0 +1,37 @@
+    /// <summary>
+    /// Details the result of attempting to encode a record to a buffer
+    /// </summary>
+    public readonly struct BebopWriteResult
```
Because the method is `EncodeIntoBuffer` this struct should be called `BebopEncodeResult`

ProgrammerAl replied:
Ended up deleting this struct.

AndrewMD5 commented on Runtime/C#/Runtime/WriteResult.cs:
```diff
@@ -0,0 +1,29 @@
+    /// <summary>
+    /// Indicates the result of writing to a buffer
+    /// </summary>
+    public enum WriteResult
```
`WriteResult` -> `EncodeResult`

AndrewMD5 reviewed (changes requested):
Thank you for the contribution! Here is some quick feedback.

ProgrammerAl commented on Runtime/C#/Runtime/WriteResult.cs:
```diff
@@ -0,0 +1,29 @@
+    /// <summary>
+    /// Indicates the result of writing to a buffer
+    /// </summary>
+    public enum WriteResult
```
Ended up deleting this enum.

ProgrammerAl commented:
@AndrewMD5 Feel free to give it another review. I changed how this is implemented. The constructor to BebopWriter takes a boolean flag to know if it's allowed to resize. If it can't, it throws an exception. I removed the struct/enum that returned a result of the write since now any return is valid.

AndrewMD5 commented on Compiler/Program.cs:
```diff
@@ -30,9 +30,17 @@ private static async Task WriteHelpText()
         {
-            Log.Formatter = CommandLineFlags.FindLogFormatter(Environment.GetCommandLineArgs());
+            var args = new[]
+            {
+                //"--lang", "cs",
+                "--namespace", "Bebop.Codegen",
+                "--files", "C:/GitHub/ProgrammerAl/bebop/Laboratory/Schemas/Valid/jazz.bop",
+                "--cs", "C:/GitHub/ProgrammerAl/bebop/Laboratory/C#/GeneratedTestCode/Output.g.cs"
+            };
+
+            Log.Formatter = CommandLineFlags.FindLogFormatter(args);
 
-            if (!CommandLineFlags.TryParse(Environment.GetCommandLineArgs(), out _flags, out var message))
+            if (!CommandLineFlags.TryParse(args, out _flags, out var message))
```
I don't think this was an intended change to commit.

ProgrammerAl replied:
Woops, it was for local testing. I'll undo that change.

AndrewMD5 commented on Runtime/C#/Runtime/BebopWriter.cs:
```diff
@@ -123,9 +144,16 @@ private void GrowBy(int amount)
+                }
+                else
+                {
+                    throw new BebopViewException("This Bebop View cannot grow the buffer.");
```
Throwing here is of course the the easiest method to fault writing, but I think in a future change I will look into how we can propagate a `-1` up the call stack to allow an opportunity for the calling application to allocate a properly sized buffer without throwing in a potential hot path.

AndrewMD5 commented:
> @AndrewMD5 Feel free to give it another review. I changed how this is implemented. The constructor to BebopWriter takes a boolean flag to know if it's allowed to resize. If it can't, it throws an exception. I removed the struct/enum that returned a result of the write since now any return is valid.

Sorry for the late review! Only one change requested and then it LGTM

AndrewMD5 commented:
Looks like test are failing @ProgrammerAl 


```
The build failed. Fix the build errors and run again.
   Compiling bebop v0.1.0 (/home/runner/work/bebop/bebop/Runtime/Rust)
   Compiling integration-testing v0.1.0 (/home/runner/work/bebop/bebop/Laboratory/Integration/Rust)
    Finished dev [unoptimized + debuginfo] target(s) in 7.73s
     Running `Rust/target/debug/examples/encode`
Testing that all implementations can decode all encodings...
rs.enc
memory allocation of 173946175504 bytes failed
cs.enc
Rust failed to decode decode c#.
cpp.enc
Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1180,33): error CS0508: 'Album.EncodeImmutably(int)': return type must be 'ImmutableArray<byte>' to match overridden member 'BaseBebopRecord.EncodeImmutably(int)' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1130,22): error CS0534: 'Album' does not implement inherited abstract member 'BaseBebopRecord.EncodeIntoBuffer(byte[])' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]

The build failed. Fix the build errors and run again.
C# decode failed.
Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1180,33): error CS0508: 'Album.EncodeImmutably(int)': return type must be 'ImmutableArray<byte>' to match overridden member 'BaseBebopRecord.EncodeImmutably(int)' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1130,22): error CS0534: 'Album' does not implement inherited abstract member 'BaseBebopRecord.EncodeIntoBuffer(byte[])' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]

The build failed. Fix the build errors and run again.
C# failed to decode rust.
Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1180,33): error CS0508: 'Album.EncodeImmutably(int)': return type must be 'ImmutableArray<byte>' to match overridden member 'BaseBebopRecord.EncodeImmutably(int)' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1130,22): error CS0534: 'Album' does not implement inherited abstract member 'BaseBebopRecord.EncodeIntoBuffer(byte[])' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]

The build failed. Fix the build errors and run again.
C# failed to decode c++.
Offset is outside the bounds of the DataView
TypeScript failed to decode c#.
terminate called after throwing an instance of 'bebop::MalformedPacketException'
  what():  malformed Bebop packet
C++ failed to decode c#
Error: Process completed with exit code 1.
```

ProgrammerAl commented:
> Looks like test are failing @ProgrammerAl
> 
> ```
> The build failed. Fix the build errors and run again.
>    Compiling bebop v0.1.0 (/home/runner/work/bebop/bebop/Runtime/Rust)
>    Compiling integration-testing v0.1.0 (/home/runner/work/bebop/bebop/Laboratory/Integration/Rust)
>     Finished dev [unoptimized + debuginfo] target(s) in 7.73s
>      Running `Rust/target/debug/examples/encode`
> Testing that all implementations can decode all encodings...
> rs.enc
> memory allocation of 173946175504 bytes failed
> cs.enc
> Rust failed to decode decode c#.
> cpp.enc
> Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1180,33): error CS0508: 'Album.EncodeImmutably(int)': return type must be 'ImmutableArray<byte>' to match overridden member 'BaseBebopRecord.EncodeImmutably(int)' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
> Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1130,22): error CS0534: 'Album' does not implement inherited abstract member 'BaseBebopRecord.EncodeIntoBuffer(byte[])' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
> 
> The build failed. Fix the build errors and run again.
> C# decode failed.
> Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1180,33): error CS0508: 'Album.EncodeImmutably(int)': return type must be 'ImmutableArray<byte>' to match overridden member 'BaseBebopRecord.EncodeImmutably(int)' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
> Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1130,22): error CS0534: 'Album' does not implement inherited abstract member 'BaseBebopRecord.EncodeIntoBuffer(byte[])' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
> 
> The build failed. Fix the build errors and run again.
> C# failed to decode rust.
> Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1180,33): error CS0508: 'Album.EncodeImmutably(int)': return type must be 'ImmutableArray<byte>' to match overridden member 'BaseBebopRecord.EncodeImmutably(int)' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
> Error: /home/runner/work/bebop/bebop/Laboratory/Integration/schema.cs(1130,22): error CS0534: 'Album' does not implement inherited abstract member 'BaseBebopRecord.EncodeIntoBuffer(byte[])' [/home/runner/work/bebop/bebop/Laboratory/Integration/IntegrationTesting.csproj]
> 
> The build failed. Fix the build errors and run again.
> C# failed to decode c++.
> Offset is outside the bounds of the DataView
> TypeScript failed to decode c#.
> terminate called after throwing an instance of 'bebop::MalformedPacketException'
>   what():  malformed Bebop packet
> C++ failed to decode c#
> Error: Process completed with exit code 1.
> ```

Looks like I didn't enable this functionality for Unions. Got that in the latest commit and added a union type to one of the schemas so this isn't missed again in the future.
